tempesta_fw/cache.c

```diff
@@ -1847,54 +1886,138 @@ tfw_cache_purge_method(TfwHttpReq *req)
+ * right before data is pushed into network.
+ * - for h2 connections - copying happens on constructing responses from
+ * cache, and SKBTX_SHARED_FRAG is left unset allowing in-place crypto
+ * operations.
+ *
+ * Since we can't encrypt shared data in-place we always copy it, so we need
+ * reserve some space in cached pages to avoid extra skb fragmentation. To do
+ * so, we reserve space for h2 frame header only before first body fragment,
+ * since it shares page with cached headers. The other body chunks begin at
+ * beginning of the cache record page right after cache record header. Since
+ * cache record header is always bigger than h2 frame header there is no need
+ * to reserve extra space for h2 frame header.
+ */
+static int
+tfw_cache_add_body_page(TfwMsgIter *it, char *p, int sz, TfwFrameHdr *frame_hdr,
+			bool h2, bool *first_frag, bool last_frag)
+{
+	int off;
+	struct page *page;
+
+	/*
+	 * @sz is guarantied to be bigger than FRAME_HEADER_SIZE if frame header
+	 * is stored in cache before data. True for first fragment.
+	 */
+	if (h2) {
+		char *new_p;
+		if (!(page = alloc_page(GFP_ATOMIC))) {
+			return -ENOMEM;
+		}
+		new_p = page_address(page);
+		off = 0;
+		frame_hdr->flags = last_frag ? HTTP2_F_END_STREAM : 0;
+		if (*first_frag) {
+			frame_hdr->length = sz - FRAME_HEADER_SIZE;
+			memcpy_fast(new_p, p, sz);
+		}
+		else {
+			frame_hdr->length = sz;
+			memcpy_fast(new_p + FRAME_HEADER_SIZE, p, sz);
+			sz += FRAME_HEADER_SIZE;
+		}
+		tfw_h2_pack_frame_header(new_p, frame_hdr);
+	}
+	else {
+		off = ((unsigned long)p & ~PAGE_MASK);
+		if (*first_frag) {
+			off += FRAME_HEADER_SIZE;
+			sz -= FRAME_HEADER_SIZE;
+		}
+		page = virt_to_page(p);
+	}
+
+	skb_fill_page_desc(it->skb, it->frag, page, off, sz);
+	skb_frag_ref(it->skb, it->frag);
+	ss_skb_adjust_data_len(it->skb, sz);
+	++it->frag;
+
+	*first_frag = false;
+
+	return 0;
+}
+
 /**
  * Build the message body as paged fragments of skb.
  * See do_tcp_sendpages() as reference.
  */
 static int
-tfw_cache_build_resp_body(TDB *db, TdbVRec *trec, TfwMsgIter *it, char *p)
+tfw_cache_build_resp_body(TDB *db, TdbVRec *trec, TfwMsgIter *it, char *p,
+			  unsigned long body_sz, bool h2, unsigned int stream_id)
 {
 	int r;
+	TfwFrameHdr frame_hdr = {.stream_id = stream_id, .type = HTTP2_DATA};
+	bool first_frag = true;
 
 	if (WARN_ON_ONCE(!it->skb_head))
 		return -EINVAL;
 	/*
 	 * If all skbs/frags are used up (see @tfw_http_msg_expand_data()),
 	 * create new skb with empty frags to reference the cached body;
-	 * otherwise, use next empty frag in current skb.
+	 * otherwise, use next empty frag in current skb. Constructing h2
+	 * responses from cache always imply data copies, for SKBTX_SHARED_FRAG
+	 * skbs, copy will happen twice. If body is relatively small keep
+	 * using existing skb and allow double copy, otherwise allocate a new
+	 * skb even if previous is not fully populated.
 	 */
-	if (!it->skb) {
+	if (!it->skb || (it->frag >= MAX_SKB_FRAGS)
```

Comment:
Now we do not increase `it->frag` here, thus it looks like in case of `(it->skb && it->frag < MAX_SKB_FRAGS)` we will overwrite the current fragment of the skb - in the `tfw_cache_add_body_page()` procedure.

Reply:
Good catch, thanks!